wolftpm/tpm2_wrap.h

```diff
@@ -2272,6 +2279,87 @@ WOLFTPM_API int wolfTPM2_CreateAndLoadAIK(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* aikKe
 */
 WOLFTPM_API int wolfTPM2_GetTime(WOLFTPM2_KEY* aikKey, GetTime_Out* getTimeOut);
 
+#ifdef WOLFTPM2_CERT_GEN
+
+WOLFTPM_API int wolfTPM2_CSR_SetCustomExt(WOLFTPM2_DEV* dev, WOLFTPM2_CSR* csr,
+    int critical, const char *oid, const byte *der, word32 derSz);
+
+WOLFTPM_API int wolfTPM2_CSR_SetKeyUsage(WOLFTPM2_DEV* dev, WOLFTPM2_CSR* csr,
+    const char* keyUsage);
+
+WOLFTPM_API int wolfTPM2_CSR_SetSubject(WOLFTPM2_DEV* dev, WOLFTPM2_CSR* csr,
+    const char* subject);
+
+WOLFTPM_API int wolfTPM2_CSR_MakeAndSign_ex(WOLFTPM2_DEV* dev, WOLFTPM2_CSR* csr,
+    WOLFTPM2_KEY* key, int outFormat, byte* out, int outSz,
+    int sigType, int selfSign, int devId);
+
+WOLFTPM_API int wolfTPM2_CSR_MakeAndSign(WOLFTPM2_DEV* dev, WOLFTPM2_CSR* csr,
+    WOLFTPM2_KEY* key, int outFormat, byte* out, int outSz);
+
+/*!
+    \ingroup wolfTPM2_Wrappers
+    \brief Helper for Certificate Signing Request (CSR) generation using a TPM based key (WOLFTPM2_KEY).
+        Single shot API for outputting a CSR or self-signed cert based on TPM key.
+
+    \return Success: Positive integer (size of the output)
+    \return TPM_RC_FAILURE: generic failure (check TPM IO and TPM return code)
+    \return BAD_FUNC_ARG: check the provided arguments
+
+    \param dev pointer to a TPM2_DEV struct
+    \param key pointer to a loaded WOLFTPM2_KEY structure
+    \param subject distinguished name string using /CN= syntax.
+        Example: "/C=US/ST=Washington/L=Seattle/O=wolfSSL/OU=Development/CN=www.wolfssl.com/emailAddress=info@wolfssl.com"
+    \param keyUsage string list of comma separated key usage attributes.
+        Possible values: any, serverAuth, clientAuth, codeSigning, emailProtection, timeStamping and OCSPSigning
+        Default: "serverAuth,clientAuth,codeSigning"
+    \param outFormat WOLFSSL_FILETYPE_ASN1 or WOLFSSL_FILETYPE_PEM
+    \param out destination buffer for CSR as ASN.1/DER or PEM
+    \param outSz destination buffer maximum size
+    \param sigType Use 0 to automatically select SHA2-256 based on keyType (CTC_SHA256wRSA or CTC_SHA256wECDSA).
+        See wolfCrypt "enum Ctc_SigType" for list of possible values.
+    \param devId The device identifier used when registering the crypto callback. Use INVALID_DEVID (-2) to
+        automatically register the required crypto callback.
+    \param selfSign If set to 1 (non-zero) then result will be a self signed certificate.
+        Zero (0) will generate a CSR to be used by a CA to sign and make into a "real" certificate.
```

Comment:
In both cases, (zero, non-zero) this is a self signed object. (CA cert or CSR) .  Can you call this parameter `is_ca_cert` or something similar? (admittedly not the best name. )